File: src/quake.py
import math
from .entity import Entity
from .sounds import Sound

class Quake(Entity):
    
    # Static options
    OPTION_QuakeDuration = 5.0
    OPTION_QuakeDropRate = 0.2
    OPTION_TimeTillFirstQuake = 90.0
    OPTION_TimeBetweenQuakes = 30.0
    OPTION_ShakeAmplitude = 0.05
    OPTION_ShakeFrequency = 50.0

    def __init__(self, game):
        super().__init__(game)
        self._earthquake = False
        self._earthquake_countdown = Quake.OPTION_TimeTillFirstQuake
        self._x = 0.0
        self._rumble = None

    def __del__(self):
        # Stop sound if any
        self._rumble = None
        # Recenter
        if self._earthquake and self._game.get_interface():
            self._game.get_interface().offset_viewport(0.0, 0.0)

    @staticmethod
    def read_settings(settings):
        Quake.OPTION_QuakeDuration = settings.get_float("Quake", "QuakeDuration", 5.0)
        Quake.OPTION_QuakeDropRate = settings.get_float("Quake", "QuakeDropRate", 0.2)
        Quake.OPTION_TimeTillFirstQuake = settings.get_float("Quake", "TimeTillFirstQuake", 90.0)
        Quake.OPTION_TimeBetweenQuakes = settings.get_float("Quake", "TimeBetweenQuakes", 30.0)
        Quake.OPTION_ShakeAmplitude = settings.get_float("Quake", "ShakeAmplitude", 0.05)
        Quake.OPTION_ShakeFrequency = settings.get_float("Quake", "ShakeFrequency", 50.0)

    def draw(self):
        # Earthquake is invisible
        pass

    def update(self, time):
        self._earthquake_countdown -= time

        if self._earthquake:
            # Drop terrain
            if self._game.get_landscape():
                self._game.get_landscape().drop_terrain(time * Quake.OPTION_QuakeDropRate)

            # Shake screen
            self._x = math.sin((Quake.OPTION_QuakeDuration - self._earthquake_countdown) * Quake.OPTION_ShakeFrequency) * Quake.OPTION_ShakeAmplitude
            
            if self._game.get_interface():
                self._game.get_interface().offset_viewport(self._x, 0.0)

            if self._earthquake_countdown < 0.0:
                self._earthquake = False
                self._earthquake_countdown = Quake.OPTION_TimeBetweenQuakes
                
                if self._game.get_interface():
                    self._game.get_interface().offset_viewport(0.0, 0.0)
                self._rumble = None # Stop sound
        
        elif self._earthquake_countdown < 0.0:
            self._earthquake = True
            self._earthquake_countdown = Quake.OPTION_QuakeDuration
            
            # Start sound
            if self._game.get_sound():
                 # 2 is hardcoded sound ID for rumble in C++ original? (quake.cc line 201)
                 # "_rumble = new cSound::cSoundSource (_game->getSound (), 2, true);"
                 self._rumble = self._game.get_sound().SoundSource(self._game.get_sound(), 2, True)

        return True

File: groundfire-0.25/src/quake.cc
////////////////////////////////////////////////////////////////////////////////
//
//               Groundfire
//
////////////////////////////////////////////////////////////////////////////////
//
// Copyright (c) 2004, Tom Russell (tom@groundfire.net)
//
// This file is part of the Groundfire project, distributed under the MIT 
// license. See the file 'COPYING', included with this distribution, for a copy
// of the full MIT licence.
//
////////////////////////////////////////////////////////////////////////////////
//
//   File name : quake.cc
//
//          By : Tom Russell
//
//        Date : 24-Feb-04
//
// Description : Handles the earthquakes that happen each round.
//
//
//
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
// Includes
////////////////////////////////////////////////////////////////////////////////
#include "quake.hh"
#include "game.hh"
#include "landscape.hh"

////////////////////////////////////////////////////////////////////////////////
// Define Static Member Variables
////////////////////////////////////////////////////////////////////////////////

float cQuake::OPTION_QuakeDuration;
float cQuake::OPTION_QuakeDropRate;
float cQuake::OPTION_TimeTillFirstQuake;
float cQuake::OPTION_TimeBetweenQuakes;
float cQuake::OPTION_ShakeAmplitude;
float cQuake::OPTION_ShakeFrequency;

////////////////////////////////////////////////////////////////////////////////
// Public Member Functions
////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////
//
// Function    : cQuake
//
// Description : Constructor
//
////////////////////////////////////////////////////////////////////////////////
cQuake::cQuake
(
    cGame * game
)
        : cEntity (game)
{
    _earthquake          = false;
    _earthquakeCountdown = OPTION_TimeTillFirstQuake;

    _x = 0.0f;

#ifndef NOSOUND
    _rumble = NULL;
#endif
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : ~cQuake
//
// Description : Destructor
//
////////////////////////////////////////////////////////////////////////////////
cQuake::~cQuake
(
)
{
#ifndef NOSOUND
    if (_rumble)
    {
        delete _rumble;
    }
#endif

    if (_earthquake)
    {
        // recentre the viewport 
        (_game->getInterface ())->offsetViewport (0.0f, 0.0f);    
    }
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : readSettings
//
// Description : Read the settings for the quake parameters
//
////////////////////////////////////////////////////////////////////////////////
void 
cQuake::readSettings
(
    cReadIniFile const & settings
)
{
    // The number of seconds the quake lasts for
    OPTION_QuakeDuration
        = settings.getFloat ("Quake", "QuakeDuration", 5.0f);

    // How quickly the terrain drops during an earthquake
    OPTION_QuakeDropRate
        = settings.getFloat ("Quake", "QuakeDropRate", 0.2f);

    // The number of seconds before the first earthquake happends
    OPTION_TimeTillFirstQuake
        = settings.getFloat ("Quake", "TimeTillFirstQuake", 90.0f);

    // The number of seconds between consecutive earthquakes
    OPTION_TimeBetweenQuakes
        = settings.getFloat ("Quake", "TimeBetweenQuakes", 30.0f);

    // The ammount by which the screen is shook
    OPTION_ShakeAmplitude
        = settings.getFloat ("Quake", "ShakeAmplitude", 0.05f);

    // The speed of the shaking
    OPTION_ShakeFrequency
        = settings.getFloat ("Quake", "ShakeFrequency", 50.0f);
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : draw
//
// Description : Draw the earthquake (earthquakes are invisible)
//
////////////////////////////////////////////////////////////////////////////////
void
cQuake::draw
(
    void
)
{
    // do nuttin'
}

////////////////////////////////////////////////////////////////////////////////
//
// Function    : update
//
// Description : Update the earthquake entity
//
////////////////////////////////////////////////////////////////////////////////
bool
cQuake::update
(
    float time
)
{
    _earthquakeCountdown -= time;

    if (_earthquake)
    {
        // The terrain collapses while an earthquake is in process.
        (_game->getLandscape ())->dropTerrain (time * OPTION_QuakeDropRate);

        // Displace the screen by '_x' units in the x-axis. 
        _x = sin ((OPTION_QuakeDuration - _earthquakeCountdown) 
                  * OPTION_ShakeFrequency) * OPTION_ShakeAmplitude;

        (_game->getInterface ())->offsetViewport (_x, 0.0f);

        if (_earthquakeCountdown < 0.0f)
        {
            // Stop the quake and set the countdown until the next quake.
            _earthquake          = false;
            _earthquakeCountdown = OPTION_TimeBetweenQuakes;

            // Reset the viewport displacement
            (_game->getInterface ())->offsetViewport (0.0f, 0.0f);

#ifndef NOSOUND
            // Stop the looping rumble sound by killing its cSoundSource object
            delete _rumble;
            _rumble = NULL;
#endif
        }
    }
    else if (_earthquakeCountdown < 0.0f)
    {
        // Start the quake and set the countdown until the quake ends.
        _earthquake          = true;
        _earthquakeCountdown = OPTION_QuakeDuration;

#ifndef NOSOUND
        // Play the looping earthquake sound
        _rumble = new cSound::cSoundSource (_game->getSound (), 2, true);
#endif
    }

    return (true);
}
